File: extension/src/tools/announcements.ts
/**
 * list_announcements — ported from src/avenue_mcp/tools/announcements.py.
 */

import { BASE_URL, avenue } from "../avenue/client.js";
import { describe, nowUtc, parseD2L } from "../avenue/dates.js";
import { asInt, isObj, pick, richText, toTextAndLinks, truncate } from "../avenue/models.js";
import { courseName } from "./context.js";

const BODY_CAP = 4000;

export async function listAnnouncements(args: {
  org_unit_id: number;
  limit?: number;
  since?: string;
}) {
  const { org_unit_id } = args;
  const limit = args.limit ?? 20;
  const now = nowUtc();
  const cutoff = args.since ? parseD2L(args.since) : null;

  const raw = await avenue.getPaged("le", `${org_unit_id}/news/`);
  const items = [];

  for (const entry of raw) {
    if (!isObj(entry)) continue;

    const posted = parseD2L(pick(entry, "StartDate", "DatePosted", "CreatedDate", "PostedDate"));
    const ends = parseD2L(pick(entry, "EndDate"));

    if (ends && ends < now) continue; // expired
    if (cutoff && posted && posted <= cutoff) continue;

    const { text, links } = toTextAndLinks(richText(pick(entry, "Body", "Content")), BASE_URL);
    const capped = truncate(text, BODY_CAP);

    items.push({
      id: asInt(pick(entry, "Id", "NewsId")),
      title: String(pick(entry, "Title", "Subject", "Name") ?? ""),
      body_text: capped.text,
      truncated: capped.truncated,
      posted_at: describe(posted),
      links,
    });
  }

  items.sort((a, b) => (b.posted_at?.utc ?? "").localeCompare(a.posted_at?.utc ?? ""));
  const limited = limit > 0 ? items.slice(0, limit) : items;

  return {
    org_unit_id,
    course_name: await courseName(org_unit_id),
    announcements: limited,
    count: limited.length,
    total_available: items.length,
    since: args.since ?? null,
  };
}

File: src/avenue_mcp/tools/announcements.py
"""list_announcements (docs/03-mcp-tools.md).

The simplest end-to-end tool, and the one that proves the pattern:
session -> client -> normalize -> structured JSON.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from avenue_mcp.client.d2l import D2LClient
from avenue_mcp.util.dates import describe, parse_d2l_date
from avenue_mcp.util.html import clean_body


async def list_announcements(
    client: D2LClient,
    *,
    org_unit_id: int,
    limit: int = 20,
    since: str | None = None,
) -> dict[str, Any]:
    """Recent course announcements, newest first."""
    tz = client.settings.timezone
    base_url = client.settings.base_url

    raw = await client.get_json("le", f"{org_unit_id}/news/")
    entries = raw if isinstance(raw, list) else []

    since_dt: datetime | None = parse_d2l_date(since) if since else None

    items: list[dict[str, Any]] = []
    for entry in entries:
        if not isinstance(entry, dict):
            continue

        posted = parse_d2l_date(entry.get("StartDate") or entry.get("CreatedDate"))
        if since_dt is not None and posted is not None and posted <= since_dt:
            continue

        # Bodies are HTML. Return clean text plus the links separately —
        # raw markup wastes context, and dropping links loses the Zoom
        # link or the reading the user actually needs.
        body_html = (entry.get("Body") or {}).get("Html") or (entry.get("Body") or {}).get("Text")
        text, links = clean_body(body_html, base_url=base_url)

        items.append(
            {
                "id": entry.get("Id"),
                "title": entry.get("Title") or "",
                "body_text": text,
                "posted_at": describe(posted, tz),
                "links": links,
                "is_pinned": bool(entry.get("IsPinned", False)),
            }
        )

    # Newest first; undated items sort last rather than crashing the compare.
    items.sort(key=lambda i: (i["posted_at"] or {}).get("utc") or "", reverse=True)
    total = len(items)

    return {
        "org_unit_id": org_unit_id,
        "announcements": items[:limit],
        "count": min(total, limit),
        "total_available": total,
    }
